File: ABC/ABC355/C/main.cc
#include <iostream>
#include <set>
#include <vector>

using namespace std;

int main() {
  int n, t;
  cin >> n >> t;

  vector<set<int>> sr(n), sc(n);
  set<int> sd1, sd2;

  for (int i = 0; i < t; ++i) {
    int a;
    cin >> a;
    int r = --a / n;
    int c = a % n;
    if ((sr[r].insert(c).second && (int)size(sr[r]) == n) ||
        (sc[c].insert(r).second && (int)size(sc[c]) == n) ||
        (r == c && sd1.insert(r).second && (int)size(sd1) == n) ||
        (r == n - 1 - c && sd2.insert(r).second && (int)size(sd2) == n)) {
      cout << i + 1 << endl;
      return 0;
    }
  }

  cout << "-1" << endl;
  return 0;
}

File: ABC/ABC355/C/main.py
n, t = map(int, input().split())
a = list(map(int, input().split()))

sr = [set() for _ in range(n)]
sc = [set() for _ in range(n)]
sd1 = set()
sd2 = set()

for i, x in enumerate(a):
    r = (x - 1) // n
    c = (x - 1) % n
    sr[r].add(c)
    sc[c].add(r)
    if r == c:
        sd1.add(r)
    if r == n - 1 - c:
        sd2.add(r)

    if max(len(sr[r]), len(sc[c]), len(sd1), len(sd2)) == n:
        print(i + 1)
        exit()

print('-1')
